src/pymor/algorithms/to_matrix.py

```diff
@@ -0,0 +1,117 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2016 pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import scipy.linalg as spla
+import scipy.sparse as sps
+import scipy.sparse.linalg as spsla
+
+from pymor.operators.block import BlockOperator
+from pymor.operators.constructions import (AdjointOperator, ComponentProjection, Concatenation, IdentityOperator,
+                                           LincombOperator, VectorArrayOperator, ZeroOperator)
+from pymor.operators.numpy import NumpyMatrixOperator
+
+
+def to_matrix(op, format=None, mu=None):
+    """Transfrom construction of NumpyMatrixOperators to NumPy or SciPy array
+
+    Parameters
+    ----------
+    op
+        Operator.
+    format
+        Format of the resulting |SciPy sparray|.
+        If `None`, a dense format is used.
+    mu
+        |Parameter|.
+
+    Returns
+    -------
+    res
+        Equivalent matrix.
+    """
+    assert format is None or format in ('bsr', 'coo', 'csc', 'csr', 'dia', 'dok', 'lil')
+    mapping = {
+        'bsr': sps.bsr_matrix,
+        'coo': sps.coo_matrix,
+        'csc': sps.csc_matrix,
+        'csr': sps.csr_matrix,
+        'dia': sps.dia_matrix,
+        'dok': sps.dok_matrix,
+        'lil': sps.lil_matrix
+    }
+    return _to_matrix(op, format, mapping, mu)
+
+
+def _to_matrix(op, format, mapping, mu):
```

Comment:
It might be good to do `op = op.assemble(mu)` as a first step. This will give you a new operator which often is already as close to a matrix as possible (take a look at the implementations of `assemble` in `operators.constructions` and `operators.numpy`). Calling `assemble` might be useful to catch some cases (think of external solvers) you are unable to handle. On the other hand, your code tries harder to convert operators to matrices than `assemble` does, so both have their justification. (I guess the semantics of `assemble` are not completely clear at the moment, but the rough meaning might be 'try to precomute as much as reasonably possible, given a fixed parameter `mu`.)